src/EditorFeatures/CSharpTest/Completion/CompletionProviders/SymbolCompletionProviderTests.cs

```diff
@@ -369,20 +369,166 @@ class CL {}";
         }
 
+        [WorkItem(7213, "https://github.com/dotnet/roslyn/issues/7213")]
         [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
-        public async Task NamespaceName1()
+        public async Task NamespaceName_Unqualified_TopLevelNoPeers()
         {
-            await VerifyItemIsAbsentAsync(AddUsingDirectives("using System;", @"namespace $$"), @"String");
-            await VerifyItemIsAbsentAsync(AddUsingDirectives("using System;", @"namespace $$"), @"System");
+            var source = @"using System;
+
+namespace $$";
+
+            await VerifyItemExistsAsync(source, "System");
+            await VerifyItemIsAbsentAsync(source, "String");
+        }
+
+        [WorkItem(7213, "https://github.com/dotnet/roslyn/issues/7213")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task NamespaceName_Unqualified_TopLevelWithPeer()
+        {
+            var source = @"
+namespace A { }
+
+namespace $$";
+
+            await VerifyItemExistsAsync(source, "A");
         }
 
+        [WorkItem(7213, "https://github.com/dotnet/roslyn/issues/7213")]
         [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
-        public async Task NamespaceName2()
+        public async Task NamespaceName_Unqualified_NestedWithNoPeers()
         {
-            await VerifyItemIsAbsentAsync(@"namespace $$", @"String");
-            await VerifyItemIsAbsentAsync(@"namespace $$", @"System");
+            var source = @"
+namespace A
+{
+    namespace $$
+}";
+
+            await VerifyNoItemsExistAsync(source);
         }
 
+        [WorkItem(7213, "https://github.com/dotnet/roslyn/issues/7213")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task NamespaceName_Unqualified_NestedWithPeer()
+        {
+            var source = @"
+namespace A
+{
+    namespace B { }
+
+    namespace $$
+}";
+
+            await VerifyItemIsAbsentAsync(source, "A");
+            await VerifyItemExistsAsync(source, "B");
+        }
+
+        [WorkItem(7213, "https://github.com/dotnet/roslyn/issues/7213")]
+        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+        public async Task NamespaceName_Unqualified_ExcludesCurrentDeclaration()
+        {
+            var source = @"namespace N$$S";
+
+            await VerifyItemIsAbsentAsync(source, "NS");
+        }
+
+//        [WorkItem(7213, "https://github.com/dotnet/roslyn/issues/7213")]
+//        [Fact, Trait(Traits.Feature, Traits.Features.Completion)]
+//        public async Task NamespaceName_Unqualified_IncompleteDeclaration()
+//        {
+//            var source = @"
+//namespace A.B.C0 { }
+//
+//namespace A
+//{
+//    namespace B
+//    {
+//        namespace C$$
+//
+//        namespace C1.X { }
+//    }
+//
+//    namespace B.C2.Y { }
+//}
+//
+//namespace A.B.C3.Z { }";
+//
+//            await VerifyItemIsAbsentAsync(source, "C0");
```

Comment:
It wouldn't, it's probably a typo. If I recall, the first line should have been `A.B.B.C0`. The reason the code to handle this got so messy is because the missing closing brace throws off the nesting for _all_ subsequent namespaces, so C2 is actually `A.B.B.C2`, C3 is `A.A.B.C3`, and so on. The intent was only add these "incorrect" namespaces to the completion list if they intersected the unbalanced region. So, if I hadn't messed it up, C0 should have been excluded.

In the end, it didn't seem worth it.